File: system-tests/http-server-shutdown/src/test/java/com/teamscale/tia/HttpServerShutdownSystemTest.java
package com.teamscale.tia;

import org.assertj.core.api.Assertions;
import org.conqat.lib.commons.io.ProcessUtils;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

import java.util.concurrent.TimeUnit;

/**
 * Runs the agent with the HTTP server enabled and makes sure that the profiled application is shut down. This ensures
 * that the HTTP server is not blocking application shutdown with non-daemon threads.
 */
public class HttpServerShutdownSystemTest {

	@Test
	@Timeout(value = 10, unit = TimeUnit.SECONDS) // if the test exceeds the timeout, the shutdown didn't succeed
	public void testShutdown() throws Exception {
		String agentJar = System.getenv("AGENT_JAR");
		String sampleJar = System.getenv("SAMPLE_JAR");
		ProcessUtils.ExecutionResult result = ProcessUtils.execute(
				new ProcessBuilder("java", "-javaagent:" + agentJar + "=http-server-port=31223", "-jar", sampleJar));
		System.out.println(result.getStderr());
		System.out.println(result.getStdout());
		Assertions.assertThat(result.getReturnCode()).isEqualTo(0);
	}

}

File: system-tests/http-server-shutdown/src/test/kotlin/com/teamscale/tia/HttpServerShutdownSystemTest.kt
package com.teamscale.tia

import com.teamscale.test.commons.SystemTestUtils
import org.assertj.core.api.Assertions.assertThat
import org.conqat.lib.commons.io.ProcessUtils
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Timeout
import java.util.concurrent.TimeUnit

/**
 * Runs the agent with the HTTP server enabled and makes sure that the profiled application is shut down. This ensures
 * that the HTTP server is not blocking application shutdown with non-daemon threads.
 */
class HttpServerShutdownSystemTest {
	@Test
	@Timeout(value = 10, unit = TimeUnit.SECONDS) // if the test exceeds the timeout, the shutdown didn't succeed
	@Throws(Exception::class)
	fun testShutdown() {
		val agentJar = System.getenv("AGENT_JAR")
		val sampleJar = System.getenv("SAMPLE_JAR")
		val result = ProcessUtils.execute(
			ProcessBuilder(
				"java",
				"-javaagent:$agentJar=http-server-port=${SystemTestUtils.AGENT_PORT}",
				"-jar", sampleJar
			)
		)
		println(result.stderr)
		println(result.stdout)
		assertThat(result.returnCode).isEqualTo(0)
	}
}
